readthedocs/doc_builder/backends/mkdocs.py

```diff
@@ -327,27 +327,60 @@ class MkdocsHTML(BaseMkdocs):
     build_dir = '_build/html'
 
 
-class SafeLoaderIgnoreUnknown(yaml.SafeLoader):  # pylint: disable=too-many-ancestors
+class ProxyPythonName(yaml.YAMLObject):
+    def __init__(self, value):
+        self.value = value
+
+
+class SafeLoader(yaml.SafeLoader):  # pylint: disable=too-many-ancestors
 
     """
-    YAML loader to ignore unknown tags.
+    Safe YAML loader.
+
+    This loader parses special ``!!python/name:`` tags without actually
+    importing or executing code. Every other special tag is ignored.
 
     Borrowed from https://stackoverflow.com/a/57121993
+    Issue https://github.com/readthedocs/readthedocs.org/issues/7461
     """
 
     def ignore_unknown(self, node):  # pylint: disable=no-self-use, unused-argument
         return None
 
+    def construct_python_name(self, suffix, node):  # pylint: disable=no-self-use, unused-argument
+        return ProxyPythonName(suffix)
+
 
-SafeLoaderIgnoreUnknown.add_constructor(None, SafeLoaderIgnoreUnknown.ignore_unknown)
+class SafeDumper(yaml.SafeDumper):
+
+    """
+    Safe YAML dumper.
+
+    This dumper allows to avoid losing values of special tags that
+    were parsed by our safe loader.
+    """
+
+    def represent_name(self, data):
+        return self.represent_scalar("tag:yaml.org,2002:python/name:" + data.value, "")
+
+
+SafeLoader.add_multi_constructor("tag:yaml.org,2002:python/name:", SafeLoader.construct_python_name)
+SafeLoader.add_constructor(None, SafeLoader.ignore_unknown)
+SafeDumper.add_representer(ProxyPythonName, SafeDumper.represent_name)
 
 
 def yaml_load_safely(content):
     """
-    Uses ``SafeLoaderIgnoreUnknown`` loader to skip unknown tags.
+    Uses ``SafeLoader`` loader to skip unknown tags.
 
-    When a YAML contains ``!!python/name:int`` it will complete ignore it an
-    return ``None`` for those fields instead of failing. We need this to avoid
-    executing random code, but still support these YAML files.
+    When a YAML contains ``!!python/name:int`` it will store the ``int``
+    suffix temporarily to be able to re-dump it later. We need this to avoid
+    executing random code, but still support these YAML files without
+    information loss.
     """
-    return yaml.load(content, Loader=SafeLoaderIgnoreUnknown)
+    return yaml.load(content, Loader=SafeLoader)
+
+
+def yaml_dump_safely(content, stream=None):
+    """Uses ``SafeDumper`` dumper to write YAML contents."""
+    return yaml.dump(content, stream=stream, Dumper=SafeDumper)
```

Comment:
I think it may be good to have a test for this one as well to be sure that we are not transforming the YAML and it looks exactly the same after `yaml_dump_safely`, right?

Reply:
True! I eventually added the `__eq__` method to `ProxyPythonName` because it made the test super easy to write :rofl: 

```python
def test_yaml_dump_safely():
    data = yaml_load_safely(content)
    assert yaml_load_safely(yaml_dump_safely(data)) == data
```

> exactly the same after yaml_dump_safely

It's not the case, the indentation can change (we don't have indentation here though), and an empty string is added at the end of `python/name` tags (see https://github.com/readthedocs/readthedocs.org/issues/7461#issuecomment-697234615), which is correct and expected, but different from the original input.

Reply:
As far as the data is the same, I'm OK with that.